torch/csrc/tensor/fake_tensor.cpp

```diff
@@ -0,0 +1,273 @@
+void serialize(SerializeBuffer& result, T value) {
+  serialize(result, std::underlying_type_t<T>(value));
+}
+
+void serialize(SerializeBuffer& result, const c10::SymInt& value) {
+  if (auto i = value.maybe_as_int()) {
+    result.push_back(0);
+    serialize(result, *i);
+  } else {
+    result.push_back(1);
+    serializePointer(result, value.toSymNodeImplUnowned());
+  }
+}
+
+void serialize(SerializeBuffer& result, const std::string& s) {
+  result.insert(result.end(), s.begin(), s.end());
+}
+
+void serialize(SerializeBuffer& result, c10::Device value) {
+  serialize(result, value.str());
+}
+
+template <typename T>
+void serialize(SerializeBuffer& result, const c10::ArrayRef<T>& value) {
+  serialize(result, value.size());
+  for (const auto& v : value) {
+    serialize(result, v);
+  }
+}
+
+struct Globals {
+  static const Globals& singleton() {
+    // Purposely leak this object!
+    static Globals* instance = new Globals();
+    return *instance;
+  }
+
+  Globals(const Globals&) = delete;
+  Globals& operator=(Globals&) = delete;
+
+ private:
+  Globals() = default;
+};
+
+const std::array<int, 4> DIM_ORDER_4{1, 3, 2, 0};
+const std::array<int, 5> DIM_ORDER_5{1, 4, 3, 2, 0};
+
+bool are_strides_like_channels_last(const at::Tensor& tensor) {
+  const auto& strides = tensor.sym_strides();
+  if (strides[1] == 0) {
+    return false;
+  }
+
+  const auto& shape = tensor.sym_sizes();
+
+  auto ndim = shape.size();
+  const int* dim_order = nullptr;
+  switch (ndim) {
+    case 4:
+      dim_order = DIM_ORDER_4.data();
+      break;
+    case 5:
+      dim_order = DIM_ORDER_5.data();
+      break;
+    default:
+      return false;
+  }
+
+  int64_t min = 0;
+  while (true) {
+    if (shape[*dim_order] == 0) {
+      return false;
+    }
+    if (strides[*dim_order] < min) {
+      return false;
+    }
+    if (*dim_order == 0 && min == strides[1]) {
+      return false;
+    }
+    min = strides[*dim_order].expect_int();
+    if (min > 1) {
+      min *= shape[*dim_order].expect_int();
+    }
+    if (*dim_order == 0) {
+      return true;
+    }
+    dim_order++;
+  }
+}
+
+c10::MemoryFormat suggest_memory_format(const at::Tensor& tensor) {
+  auto layout = tensor.layout();
+  if (layout != c10::Layout::Strided) {
+    return c10::MemoryFormat::Contiguous;
+  }
+
+  if (are_strides_like_channels_last(tensor)) {
+    // THPVariable_get_ndim
+    return tensor.dim() == 4 ? at::MemoryFormat::ChannelsLast
+                             : at::MemoryFormat::ChannelsLast3d;
+  }
+
+  return c10::MemoryFormat::Contiguous;
+}
+
+void serialize_stride(SerializeBuffer& result, const at::Tensor& tensor) {
+  auto layout = tensor.layout();
+  if (layout != c10::Layout::Strided) {
+    return;
+  }
+
+  // THPVariable_stride
+  serialize(result, tensor.sym_strides());
+}
+
+bool is_sparse_any(const at::Tensor& tensor) {
+  // return is_sparse_coo(tensor) || is_sparse_compressed(tensor);
+  // THPVariable_layout
```

Comment:
leftover ?

Reply:
I purposely left those comments in there because looking up the dispatch -> C++ function isn't always so obvious so I left it as a hint. I'll remove them.